# 📊 Task 1: Top 10 App Categories Analysis

**Author:** Vansh Sharma  

## Objective
Compare **Average Rating** and **Total Reviews** for the top 10 app categories based on installs.

## Filters Applied
- Rating ≥ 4.0
- Size ≥ 10 MB
- Last Updated Month = January

## Steps Performed
- Grouped data by Category
- Calculated Average Rating, Total Reviews, and Total Installs
- Selected Top 10 categories by installs
- Created a grouped bar chart for comparison

## Time Condition
The graph is displayed only between **3 PM IST to 5 PM IST**.

# Import Libraries

In [129]:
 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
import pytz

    

# Load DataSet `play_store_data`

In [96]:
import pandas as pd

play_store_data = pd.read_csv(
    r"C:\Users\Vansh Sharma\Downloads\Play Store Data (1).csv"
)

pd.DataFrame(play_store_data.head())

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


## Data Cleaning from 

- Removed corrupted rows from the dataset.
- Converted date columns into datetime format.
- Converted Reviews into numeric format.
- Handled missing values using forward fill and mode imputation.
- Prepared data for further analysis.

In [138]:
 # 1. Remove the corrupted row
play_store_data = play_store_data[play_store_data["Installs"] != "Free"]

# 2. Remove commas and plus signs
# play_store_data["Installs"] = (
#     play_store_data["Installs"]
#     .str.replace(",", "", regex=False)
#     .str.replace("+", "", regex=False)
# )

# # 3. Convert to integer
# play_store_data["Installs"] = play_store_data["Installs"].astype(int)

# 4. Convert Last Updated to datetime
play_store_data["Last Updated"] = pd.to_datetime(
    play_store_data["Last Updated"],
    errors="coerce"
)

# covert into int
play_store_data["Reviews"] = play_store_data["Reviews"].astype(int)


# fwd filling (Rating)
play_store_data["Rating"] = play_store_data["Rating"].ffill()
play_store_data["Current Ver"] = play_store_data["Current Ver"].ffill()
play_store_data["Android Ver"] = play_store_data["Android Ver"].ffill()
play_store_data["Type"] = play_store_data["Type"].fillna(
    play_store_data["Type"].mode()[0]
)




## Size Conversion

- Converted app size values into MB format.
- Converted KB values into MB.
- Removed non-numeric size values.
- Changed Size column into numeric format for filtering.

In [ ]:
# # covert size into mb
# play_store_data["Size"] = play_store_data["Size"].replace(
#     "Varies with device", np.nan
# )

# play_store_data.loc[
#     play_store_data["Size"].str.contains("k", na=False),
#     "Size"
# ] = (
#     play_store_data.loc[
#         play_store_data["Size"].str.contains("k", na=False),
#         "Size"
#     ]
#     .str.replace("k", "", regex=False)
#     .astype(float) / 1024
# )
# play_store_data.loc[
#     play_store_data["Size"].str.contains("M", na=False),
#     "Size"
# ] = (
#     play_store_data.loc[
#         play_store_data["Size"].str.contains("M", na=False),
#         "Size"
#     ]
#     .str.replace("M", "", regex=False)
#     .astype(float)
# )
# play_store_data["Size"] = play_store_data["Size"].astype(float)



# Fill `np.NAN` with median

In [103]:
play_store_data["Size"] = play_store_data["Size"].fillna(
    play_store_data["Size"].median()
)

# Remove Duplicates

In [106]:
play_store_data = play_store_data.drop_duplicates()

## Data Filtering & Aggregation

- Filtered apps with size ≥ 10 MB and January updates.
- Grouped data by Category.
- Calculated average rating, total reviews, and total installs.
- Removed categories with average rating below 4.0.
- Selected top 10 categories based on total installs.

In [123]:
filtered = play_store_data[
    (play_store_data["Size"] >= 10) &
    (play_store_data["Last Updated"].dt.month == 1)
]


filtered_summary = (
    filtered.groupby("Category", as_index=False)
    .agg({
        "Rating":"mean",
        "Reviews":"sum",
        "Installs":"sum"
    })
)


# Remove categories having average rating below 4
filtered_summary = filtered_summary[
    filtered_summary["Rating"] >= 4
]


# Top 10 categories by installs
filtered_summary = (
    filtered_summary
    .sort_values("Installs", ascending=False)
    .head(10)
)


filtered_summary["Rating"] = filtered_summary["Rating"].round(2)

filtered_summary

,Category,Rating,Reviews,Installs
9,FAMILY,4.19,4832660,"500,000+5,000,000+50,000,000+1,000,000+10,000+..."
17,MEDICAL,4.04,1427,"500+500+10,000+1,000+50+100+100+1,000+50,000+5..."
8,EVENTS,4.80,8,500+
26,TOOLS,4.20,409583,"50,000,000+10,000,000+5,000,000+5,000,000+100,..."
10,FINANCE,4.03,1428,"50,000+5,000+10,000+"
11,FOOD_AND_DRINK,4.42,746,"50,000+100,000+10+50,000+"
2,BOOKS_AND_REFERENCE,4.37,148,"50+1,000+1,000+"
25,SPORTS,4.47,1988575,"100,000,000+50,000+1,000+5,000+100+10,000,000+..."
22,PRODUCTIVITY,4.35,382020,"100,000,000+10,000,000+500,000+10,000,000+10,0..."
13,HEALTH_AND_FITNESS,4.10,19548,"100,000+1,000+"


## Visualization

- Created a grouped bar chart to compare Average Rating and Total Reviews.
- Used dual y-axis due to different scales of metrics.
- Added time-based visibility condition:
  - Graph displays only between 3 PM IST and 5 PM IST.
  - Hidden outside the specified time range.

In [141]:
# Current IST time
ist = pytz.timezone("Asia/Kolkata")
current_time = datetime.now(ist).time()


# Allowed time range
start_time = datetime.strptime("15:00", "%H:%M").time()
end_time = datetime.strptime("17:00", "%H:%M").time()


# Show graph only between 3 PM - 5 PM IST
if start_time <= current_time <= end_time:

    x = np.arange(len(filtered_summary["Category"]))
    width = 0.35

    fig, ax1 = plt.subplots(figsize=(14,6))

    # Average Rating
    ax1.bar(
        x - width/2,
        filtered_summary["Rating"],
        width,
        label="Average Rating"
    )

    ax1.set_xlabel("Category")
    ax1.set_ylabel("Average Rating")
    ax1.set_xticks(x)
    ax1.set_xticklabels(
        filtered_summary["Category"],
        rotation=90
    )


    # Total Reviews
    ax2 = ax1.twinx()

    ax2.bar(
        x + width/2,
        filtered_summary["Reviews"],
        width,
        label="Total Reviews"
    )

    ax2.set_ylabel("Total Reviews")


    plt.title(
        "Top 10 App Categories: Average Rating vs Total Reviews"
    )

    fig.legend(
        loc="upper right",
        bbox_to_anchor=(0.9,0.9)
    )

    plt.grid(axis="y")
    plt.show()


else:
    print("Graph is available only between 3 PM IST and 5 PM IST")

Graph is available only between 3 PM IST and 5 PM IST


# `KPIs`

In [140]:
# KPIs

total_categories = filtered_summary["Category"].nunique()

highest_install_category = filtered_summary.loc[
    filtered_summary["Installs"].idxmax(),
    "Category"
]

avg_rating = filtered_summary["Rating"].mean().round(2)

total_reviews = filtered_summary["Reviews"].sum()

total_installs = filtered_summary["Installs"].sum()

print("Total Categories:", total_categories)
print("Highest Install Category:", highest_install_category)
print("Average Rating:", avg_rating)
print("Total Reviews:", total_reviews)
total_installs = filtered_summary["Installs"].count()
print("Total Install:",total_installs)

Total Categories: 10
Highest Install Category: FAMILY
Average Rating: 4.3
Total Reviews: 7636143
Total Install: 10


## 📌 Key Performance Indicators (KPIs)

- **Total Categories Analyzed:** 10  
- **Top Performing Category by Installs:** FAMILY  
- **Average Rating of Top Categories:** 4.3 ⭐  
- **Total User Reviews Generated:** 7.63M  
- **Total Install Count Recorded:** 10  